### PR TITLE
lib: `compose` and `pipe` functions

### PR DESCRIPTION
**Pull request by kimburgess**
Provide base functions for function composition.

###### Motivation for this change
Just a couple of low level primitives that are likely useful across various modules / packages.

Rather than implementing these everywhere they're needed, it probably makes sense to include them here. I had a quick grep through the repo and could only find a single current implementation, which I've redirected back to this.

###### Things done

- [ ] Tested using sandboxicomposeng ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/servers/x11/xorg/overrides.nix [hunk 1]
```diff
@@ -3,7 +3,7 @@
 let
   inherit (args) stdenv makeWrapper;
   inherit (stdenv) lib isDarwin;
-  inherit (lib) overrideDerivation;
+  inherit (lib) overrideDerivation compose;
```
Comment (Mic92) [inline pkgs/servers/x11/xorg/overrides.nix]: where is this method used here?

Comment (KimBurgess) [reply pkgs/servers/x11/xorg/overrides.nix]: Hmmm... you're right, doesn't looks like that expression was actually in use there. I'll drop the inherit.

**Comment by KimBurgess** [comment]
I'm still super green to Nix, so apologies for any lack of perf characteristics - is there a reason for that suggested slowness in the implementation of `pipe`? From what I can see in the docs / implementation `foldl'` should be strict on the operator so evaluation should be the same as if everything was inlined + the overhead of flipping.

As for readability, I prefer pointfree (which these enable). The lack of a `(.)` operator isn't ideal, but consider the following:
```nix
pipe [ a b c d ]
``` 
vs
```nix
x: d (c (b (a x)))
```
It comes down to subjective stylistic preferences. Happy to keep it in my personal libs if it doesn't mesh with the style used through the rest of the repo.

**Comment by telent** [comment]
<two-cents owner="me">Entirely coincidentally I spent an hour or so yesterday looking for a `compose` function.  I'm not sure I'd guess the definition of `pipe` correctly, but I'd have thought function composition is up there with `map` or `fold`/`reduce` in terms of widespread functional idiom</two-cents>

**Comment by dtzWill** [comment]
I have a definition of this too, with comment saying along the lines of "surely it exists already and but I'm blind so here we go"... Haha.

Are these named similarly elsewhere? Consistently? If so helps a bit with reading.
I mean pipe seems pretty clear :).

If perf is bad maybe it should just be a builtin! :P

**Comment by KimBurgess** [comment]
`compose` should be fairly familiar for people from most language backgrounds.

`pipe` is also common terminology across [F#](https://en.wikibooks.org/wiki/F_Sharp_Programming/Higher_Order_Functions#The_.7C.3E_Operator), [OCaml](http://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#VAL%28|%3E%29), [Elixir](https://www.safaribooksonline.com/library/view/programming-elixir/9781680500530/f_0057.html), [Elm](https://edmz.org/design/2015/07/29/elm-lang-notes.html), [Julia](http://docs.julialang.org/en/release-0.4/stdlib/base/?highlight=|%3E#Base.|%3E), [Hack](https://docs.hhvm.com/hack/operators/pipe-operator), [LiveScript](http://livescript.net/#piping), the [tc39 proposal](https://github.com/tc39/proposal-pipeline-operator) and quite a few libraries for various languages, as well as UNIX pipes, although in most of these it's represented as the `|>` operator.

Happy to split these out if they're easier to consider in isolation.

Definitely agree that builtins would be optimal. If going that route, extending to `.` and `|>` operators would be beautiful, but I'm lazy and implementation here was trivial :)

**Comment by mmahut** [comment]
What is the status of this pull request?

**Comment by KimBurgess** [comment]
@edolstra expressed some valid performance concerns with `pipe`, which are likely reason enough to drop that from inclusion in this PR.

I've just had a grep through the current repo state and could find another implementation of `compose` [introduced here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix#L9). Unfortunately github code search is a little rudimentary, but from a [brief look across other nix projects](https://github.com/search?&q=%22compose%22+language%3ANix&type=Code) the occasional implementation pops up elsewhere too. With this I'd still argue that there could be some value in just having it as part of base lib.

Let me know which way you'd like to go and I can either drop this down to just compose, or close this off without the merge.

**Comment by turboMaCk** [comment]
In my opinion neither of this is useful unless it's defined as a infix or used in laguage where it's possible to write binary functions as infix (like haskells ``` syntax).

**Comment by Mic92** [comment]
Let's close this for now.

**Pull request closed without merging**